lightning/src/offers/offer.rs

```diff
@@ -0,0 +1,829 @@
+/// customer may request an `Invoice` for a specific quantity and using an amount enough to cover
+/// that quantity (i.e., at least `quantity * amount`). See [`Offer::amount`].
+///
+/// Offers may be denominated in currency other than bitcoin but are ultimately paid using the
+/// latter.
+///
+/// Through the use of [`BlindedPath`]s, offers provide recipient privacy.
+#[derive(Clone, Debug)]
+pub struct Offer {
+	// The serialized offer. Needed when creating an `InvoiceRequest` if the offer contains unknown
+	// fields.
+	bytes: Vec<u8>,
+	contents: OfferContents,
+}
+
+/// The contents of an [`Offer`], which may be shared with an `InvoiceRequest` or an `Invoice`.
+#[derive(Clone, Debug)]
+pub(crate) struct OfferContents {
+	chains: Option<Vec<ChainHash>>,
+	metadata: Option<Vec<u8>>,
+	amount: Option<Amount>,
+	description: String,
+	features: OfferFeatures,
+	absolute_expiry: Option<Duration>,
+	issuer: Option<String>,
+	paths: Option<Vec<BlindedPath>>,
+	quantity_min: Option<u64>,
+	quantity_max: Option<u64>,
+	signing_pubkey: Option<PublicKey>,
+}
+
+impl Offer {
+	// TODO: Return a slice once ChainHash has constants.
+	// - https://github.com/rust-bitcoin/rust-bitcoin/pull/1283
+	// - https://github.com/rust-bitcoin/rust-bitcoin/pull/1286
+	/// The chains that may be used when paying a requested invoice.
+	pub fn chains(&self) -> Vec<ChainHash> {
+		self.contents.chains
+			.as_ref()
+			.cloned()
+			.unwrap_or_else(|| vec![ChainHash::using_genesis_block(Network::Bitcoin)])
+	}
+
+	// TODO: Link to corresponding method in `InvoiceRequest`.
+	/// Metadata set by the originator. Useful for authentication and validating fields since it is
+	/// reflected in `invoice_request` messages along with all the other fields from the `offer`.
+	pub fn metadata(&self) -> Option<&Vec<u8>> {
+		self.contents.metadata.as_ref()
+	}
+
+	/// The minimum amount required for a successful payment of a single item.
+	pub fn amount(&self) -> Option<&Amount> {
+		self.contents.amount.as_ref()
+	}
+
+	/// A complete description of the purpose of the payment.
+	pub fn description(&self) -> &str {
+		&self.contents.description
+	}
+
+	/// Features pertaining to the offer.
+	pub fn features(&self) -> &OfferFeatures {
+		&self.contents.features
+	}
+
+	/// Duration since the Unix epoch when an invoice should no longer be requested.
+	///
+	/// If `None`, the offer does not expire.
+	pub fn absolute_expiry(&self) -> Option<Duration> {
+		self.contents.absolute_expiry
+	}
+
+	/// Whether the offer has expired.
+	#[cfg(feature = "std")]
+	pub fn is_expired(&self) -> bool {
+		match self.absolute_expiry() {
+			Some(seconds_from_epoch) => match SystemTime::UNIX_EPOCH.elapsed() {
+				Ok(elapsed) => elapsed > seconds_from_epoch,
+				Err(_) => false,
+			},
+			None => false,
+		}
+	}
+
+	/// The issuer of the offer, possibly beginning with `user@domain` or `domain`. Intended to be
+	/// displayed to the user, so should be human-readable.
```

Comment:
Good question... it is already valid UTF-8, so it would be just a matter of replacing control characters with a symbol. Would we want to do the same with `description`? I guess we could have a wrapper like `PrintableString` implementing `Display`.

Reply:
Add in https://github.com/lightningdevkit/rust-lightning/pull/1719/commits/282e49bc1e4bf6aa85b8b77f52ed5f79fca4bdf1